funsor/gaussian.py

```diff
@@ -447,9 +448,92 @@ def _eager_subs_real(self, subs, remaining_subs):
 
     def _eager_subs_affine(self, subs, remaining_subs):
-        affine = OrderedDict((k, extract_affine(v)) for k, v in subs)
-        assert affine
-        raise NotImplementedError('TODO')
+        # Extract an affine representation.
+        affine = OrderedDict()
+        for k, v in subs:
+            const, coeffs = extract_affine(v)
+            if (isinstance(const, Tensor) and
+                    all(isinstance(coeff, Tensor) for coeff, _ in coeffs.values())):
+                affine[k] = const, coeffs
+            else:
+                remaining_subs += (k, v),
+        if not affine:
+            return reflect(Subs, self, remaining_subs)
+
+        # Align integer dimensions.
+        old_int_inputs = OrderedDict((k, v) for k, v in self.inputs.items() if v.dtype != 'real')
+        tensors = [Tensor(self.info_vec, old_int_inputs),
+                   Tensor(self.precision, old_int_inputs)]
+        for const, coeffs in affine.values():
+            tensors.append(const)
+            tensors.extend(coeff for coeff, _ in coeffs.values())
+        new_int_inputs, tensors = align_tensors(*tensors, expand=True)
+        tensors = (Tensor(x, new_int_inputs) for x in tensors)
+        old_info_vec = next(tensors).data
+        old_precision = next(tensors).data
+        for old_k, (const, coeffs) in affine.items():
+            const = next(tensors)
+            for new_k, (coeff, eqn) in coeffs.items():
+                coeff = next(tensors)
+                coeffs[new_k] = coeff, eqn
+            affine[old_k] = const, coeffs
+        batch_shape = old_info_vec.data.shape[:-1]
+
+        # Align real dimensions.
+        old_real_inputs = OrderedDict((k, v) for k, v in self.inputs.items() if v.dtype == 'real')
+        new_real_inputs = old_real_inputs.copy()
+        for old_k, (const, coeffs) in affine.items():
+            del new_real_inputs[old_k]
+            for new_k, (coeff, eqn) in coeffs.items():
+                new_dim = len(eqn.split('->')[0].split(',')[1])
+                new_real_inputs[new_k] = reals(*coeff.shape[:new_dim])
+        old_offsets, old_dim = _compute_offsets(old_real_inputs)
+        new_offsets, new_dim = _compute_offsets(new_real_inputs)
+        new_inputs = new_int_inputs.copy()
+        new_inputs.update(new_real_inputs)
+
+        # Construct a blockwise affine representation of the substitution.
+        subs_vector = BlockVector(batch_shape + (old_dim,))
+        subs_matrix = BlockMatrix(batch_shape + (new_dim, old_dim))
+        for old_k, old_offset in old_offsets.items():
+            old_size = old_real_inputs[old_k].num_elements
+            old_slice = slice(old_offset, old_offset + old_size)
+            if old_k in new_real_inputs:
+                new_offset = new_offsets[old_k]
+                new_slice = slice(new_offset, new_offset + old_size)
+                subs_matrix[..., new_slice, old_slice] = \
+                    torch.eye(old_size).expand(batch_shape + (-1, -1))
```

Comment:
This more advanced slicing isn't covered by existing `BlockMatrix` tests in `test_gaussian.py`, right?  Would it make sense to add a couple of test cases there?

Reply:
Will do.

Reply:
This advanced indexing was already covered, but the particular sparse sequence of entries was not, so I added a test for that.